workflows/pipe-common/scripts/autoscale_sge.py

```diff
@@ -1300,10 +1311,118 @@ def _validate_existence(self, host, hosts_dict):
+        Logger.info('Finish tags processing step at %s.' % self.clock.now())
+
+    def _run_is_active(self, timestamp):
+        return timestamp > self.clock.now() - self.run_activation_timeout
+
+    def _tag_run(self, host, timestamp, last_monitored_timestamps):
+        run_id = self.common_utils.get_run_id_from_host(host)
+        self._add_worker_tag(run_id)
+        last_monitored_timestamps.last_action_timestamp = timestamp
+        last_monitored_timestamps.last_tag_timestamp = self.clock.now()
+        self.last_monitored_hosts.update({host: last_monitored_timestamps})
+
+    def _untag_run(self, host, timestamp=None, last_monitored_timestamps=None):
+        run_id = self.common_utils.get_run_id_from_host(host)
+        self._remove_worker_tag(run_id)
+        if not last_monitored_timestamps:
+            del self.last_monitored_hosts[host]
+            return
+        last_monitored_timestamps.last_action_timestamp = timestamp
+        last_monitored_timestamps.last_tag_timestamp = None
+        self.last_monitored_hosts.update({host: last_monitored_timestamps})
+
+    def _process_current_hosts(self, current_hosts, hosts_activity):
+        Logger.info("Started tags processing for running hosts '%s' ." % current_hosts)
+        for current_host in current_hosts:
+            timestamp = hosts_activity[current_host]
+            if not timestamp:
+                continue
+            last_monitored_timestamps = self.last_monitored_hosts.get(current_host, None)
+            if not last_monitored_timestamps:
+                Logger.info("Found not monitored host '%s'." % current_host)
+                last_action_marker = LastActionMarker()
+                self.last_monitored_hosts.update({current_host: last_action_marker})
+                continue
+            if self._run_is_active(timestamp):
+                if not last_monitored_timestamps.last_tag_timestamp:
+                    Logger.info("Host '%s' is active. Adding a new tag to the corresponding run." % current_host)
+                    self._tag_run(current_host, timestamp, last_monitored_timestamps)
+                    Logger.info("Added tag to run for host '%s'." % current_host)
+                # do nothing if active run was already tagged
+                continue
+            Logger.info("Host '%s' is not active." % current_host)
+            if last_monitored_timestamps and last_monitored_timestamps.last_tag_timestamp:
+                Logger.info("Removing tag from run.")
+                self._untag_run(current_host, timestamp, last_monitored_timestamps)
+                Logger.info("Removed tag from run for host %s." % current_host)
+
+    def _process_outdated_hosts(self, monitored_hosts, current_hosts):
+        for monitored_host in monitored_hosts:
+            if monitored_host not in current_hosts:
+                Logger.info("Host %s scaled down. Removing tags from run." % monitored_host)
+                self._untag_run(monitored_host)
```

Comment:
Though I do appreciate having lots of logs, at the moment I would suggest to reduce the amount of logs to only two records (add/remove tag). In the current logging setup we don't have a proper usage of logging levels (debug in specific). This log is extremely important and we shall keep it brief and human readable even for clusters even with hundreds of workers.

In the future I want to use _pipeline.log.logger.CloudPipelineLogger_ which does better job in logging level management. Once it is used, we can add more debug logs. But for now let's keep it brief.

```
...
Logger.info('Marking active host %s...' % current_host)
...
Logger.info('Marking inactive host %s...' % current_host)
```